python/graphstorm/graphpartitioning/metadata_schema.py

```diff
@@ -0,0 +1,730 @@
+            if isinstance(count, int) and count > 0:
+                continue
+            raise ValueError("Invalid value for No. of nodes per type")
+
+        prefix_sum = np.cumsum([0] + etype_counts)
+        starts = prefix_sum[:-1]
+        ends = prefix_sum[1:]
+        ranges = zip(starts, ends)
+
+        self._global_eid_offsets = dict(zip(etypes, ranges))
+
+    def _init_global_nid_offsets(self):
+        """Function to initialize the global_node_id offsets for all the
+        node types present in the input graph.
+
+        global_node_ids, or global_nids, are unique ids assigned to each
+        node of the input graph. These ids are expected to be unique ids
+        across the input graph.
+
+        The range of global_node_ids for each node type in the graph are
+        stored in a dictionary where key-value pairs are ntypes and a tuple
+        which indicates the starting and ending value (exclusive) of the
+        global_node_ids. For instance,
+            ntype1: [a, b) are the key-value pairs.
+            where ntype1 is a string describing the node type
+            and [a, b) is the range of global_node_ids for the node type
+            `ntype1`.
+        """
+        ntypes = self._data[STR_NODE_TYPE]
+        ntype_counts = self._data[STR_NUM_NODES_PER_TYPE]
+        for count in ntype_counts:
+            if isinstance(count, int) and count > 0:
+                continue
+            raise ValueError("Invalid value for No. of nodes per type")
+
+        prefix_sum = np.cumsum([0] + ntype_counts)
+        starts = prefix_sum[:-1]
+        ends = prefix_sum[1:]
+        ranges = zip(starts, ends)
+        self._global_nid_offsets = dict(zip(ntypes, ranges))
+
+    def _init_etype_maps(self):
+        """Initialization function to build unique id maps for node types of
+        the input graph. This function creates a list of edge types present in
+        the graph, and unique-id <-> edge type and reverse maps.
+        """
+        self._etypes = self._data[STR_EDGE_TYPE]
+        self._etype_id_map = {
+            etype_name: idx for idx, etype_name in enumerate(self._etypes)
+        }
+        self._id_etype_map = dict(enumerate(self._etypes))
+
+    def _init_etype_features(self):
+        """Initialization function to build a map in which key are edge types
+        in the graph and values are a list of features for the corresponding
+        edge type, if present. For instance,
+            etype1: [feature1, feature2, ....] where
+            etype1 is a edge type from the graph and
+            [feature1, feature2, ...] is a list of features associated with
+            the etype1 edge type.
+        """
+        self._etype_features = {}
+        edge_data = self._data.get(STR_EDGE_DATA, {})
+        for etype in self._data[STR_EDGE_TYPE]:
+            features = edge_data.get(etype, {})
+            if len(features) > 0:
+                self._etype_features[etype] = list(features.keys())
```

Comment:
the way that the code is written is weird. isn't this
```
if etype in edge_data:
    ....
```

Reply:
also, why do you need `self._etype_features`? isn't `edge_data` used for this purpose?

Reply:
This has to do with the user experience. Following is the reason:
1. "edge_type", list of etypes in the graph, is the used to extract features from "edge_data". This is used to filter out any erroneous etypes which may show up in the "edge_data". If we read "etypes" and "features" from "edge_data" then we might run into erroneous cases where a certain etype is present in the "edge_data" but NOT present in the "edge_type" field. We are avoiding such a scenario and instead of throwing some wierd errors in such case we are gracefully continuing with the pipeline execution -- in this case by ignoring etypes/features which might be present in the "edge_data" but not present in the "edge_type".  This ensures that the user experience is enhanced. 

"self._etype_features" contains a list of features for each valid etype in the input dataset. Because of the above reason we cannot rely on the correctness of the input dataset.

By building self._etype_features, we are abstracting away the application to figure out which features exist for a given etype. So that way any changes to "edge_data" will result only in the changes in this file - as opposed to any changes in the application itself.